Improve build (alpine and proper config with env vars)

esbite opened the pull request:
Here is the `lite` branch but rebased on the latest master and with some added improvements.

- Uses the HMD_DB_URL also inside .sequelizerc for proper database configuration using only ENV vars.
- Upgraded docker-compose.yml to version 2 and build the subfolder hackmd instead of static image.
- Latest hackmd master has renamed build script to "build", this Dockerfile respects that.
- Makes a git clone of source without history for slightly faster build time.
- Emptied out config.json, using defaults instead and just overriding a few with environment variables in the Dockerfile (NODE_ENV , HMD_DB_URL, HMD_URL_ADDPORT). Other users can still base another Dockerfile on this one and add their custom config.json if they so wish.

Here at @erasys we are currently running this in our Marathon/Mesos cluster by only setting a few environment variables to configure it. Really awesome!

jackycute commented:
Thanks a lot @esbite 😄 
We might take some to review this PR.

But first, as we collaborated with @SISheogorath before.
The lite version not able to execute `phantom.js` because it's not prebuilt for alpine.
Or if it's too hard to solve it in this PR then I will open an issue for that.

SISheogorath commented on hackmd/docker-entrypoint.sh:
```diff
@@ -1,12 +1,11 @@
 
 # run
-NODE_ENV='production' node app.js
+exec node app.js
```
We should avoid using `exec` here.

`exec` avoid the creation of a subshell. (wonderful in speech of shell programming)

But it also avoid the creation of a new PID. In case of docker that's a bit annoying and possibly a security problem. (If you hear some security experts in the forest ;))

Reason: In the container the CMD is called as PID1. PID 1 is not really random... As we know random PIDs help security -> https://security.stackexchange.com/questions/88692/do-randomized-pids-bring-more-security

So in this case I would like a subprocess. And in general it's not a problem. (There are good reasons for the PID 1 -> https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/ as initial PID)

esbite replied:
Oh cool, I had no idea about this, will change back.

jackycute commented on hackmd/config.json:
```diff
@@ -1,14 +1 @@
-        }
-    }
-}
+{"docker": {}}
```
We need at least one `config.json` that contains object with key name which corresponding to current `NODE_ENV` like `"development:" {}`

esbite replied:
Yes I tried deleting it and it didn't work. Would be cool if it would work tho. And it would also be really nice if we could drop the requirement of the `config.js` file, and just have the app use all defaults and ENV vars if no config files are set.

jackycute replied:
@esbite  I think you mean `config.json` right?

esbite replied:
@jackycute No the `config.js`, but I saw some commits in the latest `master` regarding that, maybe you have already removed it?

jackycute replied:
Oops, missed that.
Yeah, we will remove that `config.js` in the upcoming release.

SISheogorath commented on hackmd/Dockerfile:
```diff
@@ -1,34 +1,43 @@
+USER hackmd
+
+# Set some default config variables
+ENV NODE_ENV=docker
```
See (for example): http://expressjs.com/en/advanced/best-practice-performance.html#in-environment

There is more ;)

esbite replied:
Ah I see, will change back! :)

esbite commented on hackmd/Dockerfile:
```diff
@@ -1,34 +1,43 @@
-	wget -O - http://apt.postgresql.org/pub/repos/apt/ACCC4CF8.asc | apt-key add - && \
-	apt-get update && \
-	apt-get install -y git postgresql-client-9.6 build-essential
+FROM node:6-alpine
```
Okay! Will change.

esbite commented on hackmd/Dockerfile:
```diff
@@ -1,34 +1,43 @@
+RUN apk update && \
+    apk add --no-cache build-base python git && \
+    # Clone the source
+    git clone --depth 1 --branch $SourceVersion $SourceUrl /hackmd && \
```
Good suggestion. Will change.

esbite commented on hackmd/Dockerfile:
```diff
@@ -1,34 +1,43 @@
+    cd /hackmd && \
+    npm install && \
+    # Build project! (Older versions uses npm script "build:prod", newer version uses just "build")
+    build_script_name="$(if grep -q 'build:prod' package.json; then echo 'build:prod'; else echo 'build'; fi)" && \
```
We are still building 0.5.0 since I had quite a few issues trying to build master. Will keep this here until a new public hackmd release is made?

jackycute replied:
We still need to build in dockerfile, the build is for bundling the frontend resources.
And the we've changed the `build:prod` to simply `build`.

esbite commented on hackmd/Dockerfile:
```diff
@@ -1,34 +1,43 @@
+
+# Set some default config variables
+ENV NODE_ENV=docker
+ENV HMD_DB_URL=postgres://hackmd:hackmdpass@postgres:5432/hackmd
```
Considering how HackMD currently handles the options I thought it was a good idea to specify the most basic defaults needed here, makes it very obvious what ENV variables you have to set when running the container. 

They are still easily overridable by just providing an environment variable with the same name to `docker run` and then that value will just be passed into the container instead :)

jackycute replied:
I think it's better to leave all defaults to https://github.com/hackmdio/docker-hackmd/blob/lite/hackmd/config.json.
And let user override via env vars to get what they need.

esbite commented on hackmd/docker-entrypoint.sh:
```diff
@@ -1,12 +1,11 @@
-then
+if [ -f .sequelizerc ]; then
     node_modules/.bin/sequelize db:migrate
 fi
```
When writing scripts I would argue for keeping the readability that if..then offers :)

SISheogorath replied:
I an live with the syntax but I would still suggest to use `-e` instead of `-f` because it allows you to use a symlink. Maybe not needed in general but it simply is more generic.

SISheogorath replied:
Eh simply idea: As we copy the `.sequalizerc` to the image we are sure it exists. And it's never removed. There is no need to check of existence. We can simply remove the whole if-clause.

I don't the any reason to keep the database schema upgrade optional.

esbite commented on docker-compose.yml:
```diff
@@ -1,15 +1,20 @@
+      - POSTGRES_PASSWORD=hackmdpass
+      - POSTGRES_DB=hackmd
+  hackmd:
+    build: ./hackmd
```
But this is the docker building repo, at least for me who are used to building images it makes total sense that when this repository actually contains the Dockerfile specification that it should also build itself. 

I guess it might be another case for someone who only grabs the docker-compose.yml from here and expects that to work on its own. Depends on what use case you expect this repo to have.

jackycute replied:
We should use image here.
The building config is in release branch.

esbite replied:
I guess you are right, I also had lots of random issues with NPM and Webpack when building, every other build would fail for some random js file missing in the end result.

esbite replied:
I think we should keep this here tho for now, because with these changes you guys need to rebuild the image, push a new tag and update the reference here. Because some changes (like the HMD_DB_URL in sequilizerc) will not work with the old image.

jackycute replied:
Yes.

esbite replied:
💡  Tip: If you want to use this build setup to try out your Dockerfile changes, just do `docker-compose up --build` and it will rebuild the image every time you change the Dockerfile :)

SISheogorath replied:
I would suggest the following compromise: Add build to the instructions but comment it out. In case someone want to build the image he can easy uncomment it and everything wild build like it should. But in general the image is used.

esbite commented on docker-compose.yml:
```diff
@@ -1,15 +1,20 @@
+      - "3000:3000"
+    environment:
+      - HMD_DB_URL=postgres://hackmd:hackmdpass@postgres:5432/hackmd
+      - HMD_DOMAIN=localhost
```
Will it default to localhost?

jackycute replied:
No need to set above env vars.
They already set in https://github.com/hackmdio/docker-hackmd/blob/lite/hackmd/config.json

jackycute replied:
default domain will be `''`, and it still works.

esbite replied:
No, as will you see below in this pull request I completely cleared out the config.json and are opting to instead only set settings via environment variables. Much more flexible in a Docker context.

jackycute replied:
Hmm, maybe we could consider that.

esbite commented on docker-compose.yml:
```diff
@@ -1,15 +1,20 @@
+      - HMD_ALLOW_ANONYMOUS=true
+      - HMD_ALLOW_FREEURL=true
+      - HMD_EMAIL=false
+      - HMD_ALLOW_EMAIL_REGISTER=false
```
Yes this was all just the setup we are using in house.

esbite commented on docker-compose.yml:
```diff
@@ -1,15 +1,20 @@
+      - HMD_ALLOW_FREEURL=true
+      - HMD_EMAIL=false
+      - HMD_ALLOW_EMAIL_REGISTER=false
+      - HMD_IMAGE_UPLOAD_TYPE=filesystem
```
Yes this was all just the setup we are using in house.

esbite replied:
Yes this was just the setup we are using in house. I would argue tho that defaulting to uploading any image to imgur seems kinda creepy to me. When running this docker-compose.yml I would expect all of my changes to live inside my local environment and nothing should by default leak outside.

Maybe we can just define a volume for the uploads folder instead in docker-compose.yml?

jackycute commented:
OK, so I think it's better to set the defaults to the `config.json` file.
Because it will leave the env vars to the users to make their own decisions.

esbite commented:
Thank you for the feedback! I pushed some quick changes right away :)

What we still need to decide or work upon:
- Find a solution to get phantom.js PDF file generation running on Alpine.
- What image upload to use in docker-compose.yml.
- Where to put default settings, config.json or ENV statements.

## Image upload
To me it makes sense to use the filesystem image upload by default, and just define a volume where they will be stored persistently. After all this docker-compose.yml is just an example usage, and I assume that everyone setting it up will change it to their needs. 

(I was actually quite scared when I tried it out the first time and mistakenly uploaded a random image, realised it ended up on imgur and found no way to remove it from their service.)

I pushed this volume change now and it seems to work fine. Actually it is the exact same way that the default postgres database container stores its data, in a volume. I think this makes sense to have all data locally in docker by default and let the service run without any third party services by default.

But I am not heavily opinionated on this topic, we will override it in our use case anyway :)

## Configuration settings
Okay after writing like a three-page argument here about using ENV statements to set the few default values that the docker image we build actually needs to run properly, I realised that if we set settings like this they can never be overridden using a custom config.json :facepalm:

I still think that having the `NODE_ENV` here makes good sense. This was previously defined in the docker-entrypoint.sh, and as such not possible to override by any means. Now that you told me this is something that nodejs requires to optimise things properly I understand why you put it there, but I still think this should be up to the user of the image to decide.

I will then remove the ENV settings for HMD_DB_URL and HMD_URL_ADDPORT. I mean I'm not even sure they are needed anyway? I just built an image without them and seems to work fine.

FYI: If someone else uses the docker image and wants to override settings using a config.json instead of sending environment variables to docker, they can just mount a config manually and it will be used :)
```
docker run -v $(pwd)/config.json:/hackmd/config.json hackmdio/hackmd
```

esbite commented:
Getting PhantomJS to run on Alpine seems to be quite an undertaking... https://github.com/ariya/phantomjs/issues/14186
https://github.com/ariya/phantomjs/issues/14265

SISheogorath commented:
I understand your argumentation about the local image uploads. I personally would prefer it if our docker image simply works out of the box as stand alone (without the need of docker compose). I know that would be easily possible by using sqlite as default database backend in the dockerfile. In this case I would agree completely agree with the local filesystem as storage for images.

But as far as I can see we try to stay stateless with this container. So local image upload is not a real option. 

For the custom `config.json` I see no reason for someone to use it as all options are available as environment variables. Is there anyone using it? 😕 

In order to fix the PhantomJS problem you can checkout https://github.com/SISheogorath/docker-hackmd/tree/fix/pdfGenerator

It already works basically. But it still has some bugs.

esbite commented:
I see your point about "works out of the box", but setting the image upload to `filesystem` in the `docker-compose.yml` does not change the image itself. That still uses the `imgur` service, as the default settings in the hackmd codebase specify.

Getting the database to work out of the box is certainly more problematic. Setting it to use SQLite inside the image would give it the state you don't want. This might be a separate discussion.

Regarding custom `config.json` I can still see some missing environment variables, e.g. the S3 settings cannot be altered. And maybe someone would just prefer to have their settings in a JSON. Nothing in the current image would stop them from doing that :)

Neat fix with the PhantomJS! Can we rely on that pre built image to stay available for download? I guess there is no better fix at the moment, unless someone gets an official package into Alpine :)

esbite commented:
So I tried the PhantomJS stuff, it installed correctly but when trying it out it still doesn't work for me at all, just getting empty PDFs without any errors logged in docker. Since that feature is not a requirement for us I will leave that for you guys to figure out 🤔 

@SISheogorath Should I change this merge request to point to the `lite` branch then? :)

esbite commented:
I removed the IF statement around the sequelize command as you requested. And also added a comment about how to build the current Dockerfile in the docker-compose.yml.

Also tried building again using the `master` branch on hackmd, and that worked fine this time! So I updated the default SourceVersion to be `master` :)

SISheogorath commented:
I really like it. The only thing we have to do is redirect the Pull Request to the `lite` branch instead of master as long we have no working fix for the PhantomJS problem :/

I guess @jackycute agrees with that.

jackycute commented:
Yes, this seems good now.
@esbite Please send PR to `lite` branch, I would love to merge it.
Thank you guys so much 👍

esbite commented:
Merging this directly into your `lite` branch will of course cause a lot of merge conflicts. Because this branch includes the lite-commits but also is rebased on top of your latest `master` (which includes updates to 0.5.0).

![screen shot 2017-02-08 at 13 00 08](https://cloud.githubusercontent.com/assets/219730/22736492/a7bb6f94-edfe-11e6-93f5-34c81c31e956.png)

I'm kinda new to GitHub collaboration, but when we have a situation like this in our company we just force push the new rebased branch containing all commits. How do you normally solve a situation like this here? We want to keep the `lite` branch up to date with the `master` branch right? :)

I could also squash some of my review commits on this branch to make it a bit cleaner.

jackycute commented:
I would like to call out @SISheogorath since the `lite` was from him 😄

SISheogorath commented:
I guess you cherry-picked the commits. Using `git merge -s theirs <branch>` should resolve all conflicts. after resolving it you have to `git push <upstreamremote> lite` it.

It's an improved version of the lite branch so I'm fine with it :D

esbite commented:
Hi again. So I did not fully understand your last comment, but I cherry picked all my commits on top of your lite branch instead, and also merged master into it to get those latest changes. So now it can be merged without conflicts. Hopefully this looks good to you :)

SISheogorath commented:
It was more for @jackycute :D But as you unbroke it now, it looks nice. Waiting for merge \o/

jackycute commented:
Thanks @esbite @SISheogorath 
Time to merge this.
